trpcultivate_phenotypes/src/Service/TripalCultivatePhenotypesTraitsService.php

```diff
@@ -255,163 +292,289 @@ public function insertTrait($trait, $schema = NULL) {
+   *   and was not set a data type.
+   *
+   * @dependencies
+   *   getPhenoCvterm()
    */
-  public function getMethodUnit($method_id) {
-    // Configuration settings of the genus.
-    $genus_config = $this->config;
-
-    if (!isset($genus_config['unit']['id']) && $this->terms['unit_to_method_relationship_type'] <= 0) {
-      // Not configured genus and term.
+  public function getMethodUnitDataType($unit) {
+    $unit_rec = $this->getPhenoCvterm($unit, 'unit');
+    if (!$unit_rec) {
+      // Unit was not found.
       return 0;
     }
 
-    $unit = 0;
+    // Inspect the relationship table where unit has a unit - data type relationship.
+    $sql = "SELECT value FROM {1:cvtermprop} WHERE cvterm_id = :c_id AND type_id = :t_id";
 
-    // Get unit.
-    $sql = "SELECT object_id FROM {1:cvterm_relationship} WHERE subject_id = :s_id AND type_id = :t_id";
-
-    // Query values.
     $args = [
-      ':s_id' => (int) $method_id,
-      ':t_id' => $this->terms['unit_to_method_relationship_type']
+      ':c_id' => $unit_rec->cvterm_id,
+      ':t_id' => $this->terms['unit_type']
     ];
 
-    // Query unit.
-    $unit_id = $this->chado->query($sql, $args)
-      ->fetchField();
+    $data_type = $this->chado->query($sql, $args)
+      ->fetchCol();
 
-    if ($unit_id) {
-      $sql = "SELECT * FROM {1:cvterm} WHERE cvterm_id = :id AND cv_id = :c_id LIMIT 1";
-      $unit = $this->chado->query($sql, [':id' => $unit_id, ':c_id' => $genus_config['unit']['id']])
-        ->fetchObject();
+    if (count($data_type) > 1) {
+      // Unit appears to have multiple data types.
+      throw new \Exception(t('A multiple data type error occurred while retrieving a unit data type. 
+        Failed to retrieve data type for unit : @unit in cv : @cv. Multiple data types found for the same unit.',
+        ['@unit' => $unit, '@cv' => $genus_config['unit']['name']]
+      ));
     }
 
-    return $unit;
+    return ($data_type) ? reset($data_type) : 0;
   }
 
   /**
-   * Get Trait Method Unit data type.
+   * Get trait, method and unit combination.
    *
-   * @param integer $unit_id
-   *   Method unit id number (unit cvterm id).
+   * @param string|int $trait
+   *   A string value is the trait name, whereas an integer value is the trait id number.
+   * @param string|int $method
+   *   A string value is the trait method short name, whereas an integer value is the method id number.
+   * @param string|int $unit
+   *   A string value is the method unit name, whereas an integer value is the unit id number.
    *
-   * @return string
-   *   The data type of the unit, either quantitative or qualitative. False if not set.
+   * @return array
+   *   An associative array where the keys are trait, method and unit and the values
+   *   are the cvterm records for each key.
+   *
+   *   0 if any one of trait, method or unit did not return any record.
```

Comment:
Needs to be updated since it currently returns an empty array, not 0. See below comment!

Reply:
Updated doc blocks to be very clear about return value and exceptions thrown.